File: Python/oop.py
from abc import ABC, abstractmethod

# abstract class
class Product(ABC):
    def __init__(self, name, price):
        self.name = name
        self.price = price
    
    @abstractmethod
    def get_discounted_price(self):
        pass

      
# inheritance 
class Clothing(Product):
    def __init__(self, name, price, size, color):
        super().__init__(name, price)
        self.size = size
        self.color = color
    
    def get_discounted_price(self):
        if self.size == 'S':
            return self.price * 0.9
        else:
            return self.price * 0.8
          
          
# interface
class PaymentMethod(ABC):
    @abstractmethod
    def make_payment(self, amount):
        pass

      
# implementation
class CreditCard(PaymentMethod):
    def __init__(self, card_number, expiry_date):
        self.card_number = card_number
        self.expiry_date = expiry_date
    
    def make_payment(self, amount):
        print(f"Processing credit card payment for ${amount}")

File: Java/OOP.java
// abstract class 
abstract class Product {
    private String name;
    private double price;

    public Product(String name, double price) {
        this.name = name;
        this.price = price;
    }

    public String getName() {
        return name;
    }

    public double getPrice() {
        return price;
    }

    public abstract double getDiscountedPrice();
}


// inheritance
class Clothing extends Product {
    private String size;
    private String color;

    public Clothing(String name, double price, String size, String color) {
        super(name, price);
        this.size = size;
        this.color = color;
    }

    public double getDiscountedPrice() {
        if (size.equals("S")) {
            return getPrice() * 0.9;
        } else {
            return getPrice() * 0.8;
        }
    }
}

// interface
interface PaymentMethod {
    void makePayment(double amount);
}


// implementation
class CreditCard implements PaymentMethod {
    private String cardNumber;
    private String expiryDate;

    public CreditCard(String cardNumber, String expiryDate) {
        this.cardNumber = cardNumber;
        this.expiryDate = expiryDate;
    }

    public void makePayment(double amount) {
        System.out.printf("Processing credit card payment for $%.2f\n", amount);
    }
}
